Add limited support for map[string]map[string]xxx

jbrette opened the pull request:
The goal of this PR is to get openapi-gen to handle map[string]map[string]xxx in a similar way controller-gen supports it for the schema definition within the CRD.yaml and for the deep copy operation.
 
with this PR, openapi-gen start to support the construct:

```bash
openapi-gen -i "k8s.io/apimachinery/pkg/apis/meta/v1,github.com/keleustes/armada-crd/pkg/apis/armada/v1alpha1"   -o pkg   -p generated   -O openapi_generated   -r ./swagger/golden.report
```

```go
                                        "labels": {
                                                SchemaProps: spec.SchemaProps{
                                                        Description: "labels contains tbd",
                                                        Type:        []string{"object"},
                                                        AdditionalProperties: &spec.SchemaOrBool{
                                                                Allows: true,
                                                                Schema: &spec.Schema{
                                                                        SchemaProps: spec.SchemaProps{
                                                                                // limited support by openapi-gen
                                                                                //    name: map[string]github.com/keleustes/armada-crd/pkg/apis/armada/v1alpha1.ArmadaMapString
                                                                                //    type: map[string]map[string]string
                                                                        },
                                                                },
                                                        },
                                                },
                                        },

```
controller-gen seems to be able to deal with map[string]map[string]string construct

When generating the CRD.yaml with schema:
```bash
GO111MODULE=on controller-gen crd paths=./pkg/apis/armada/... crd:trivialVersions=true output:crd:dir=./kubectl output:none
```

```yaml
                labels:
                  additionalProperties:
                    additionalProperties:
                      type: string
                    type: object
                  description: labels contains tbd
                  type: object
```

controller-gen is able to deal deep-copy

```bash
GO111MODULE=on controller-gen object paths=./pkg/apis/armada/... output:object:dir=./pkg/apis/armada/v1alpha1 output:none
```

```go
        if in.Labels != nil {
                in, out := &in.Labels, &out.Labels
                *out = new(map[string]ArmadaMapString)
                if **in != nil {
                        in, out := *in, *out
                        *out = make(map[string]ArmadaMapString, len(*in))
                        for key, val := range *in {
                                var outVal map[string]string
                                if val == nil {
                                        (*out)[key] = nil
                                } else {
                                        in, out := &val, &outVal
                                        *out = make(ArmadaMapString, len(*in))
                                        for key, val := range *in {
                                                (*out)[key] = val
                                        }
                                }
                                (*out)[key] = outVal
                        }
                }
        }
```

jbrette commented:
/cc @ianhowell

jbrette commented:
/assign @lavalamp

sttts commented on pkg/generators/openapi.go:
```diff
@@ -648,13 +648,29 @@ func (g openAPITypeWriter) generateMapProperty(t *types.Type) error {
 	return nil
 }
 
+func (g openAPITypeWriter) generateSubMapLimitedSupport(parentName types.Name, t *types.Type) error {
```
what would the generic solution look like?

jbrette replied:
@stts You are right. Solution looks more generic now.

sttts commented on pkg/generators/openapi.go:
```diff
@@ -647,14 +647,41 @@ func (g openAPITypeWriter) generateMapProperty(t *types.Type) error {
+			return err
+		}
+	case types.Map:
+		if err := g.generateNestedMapProperty(t.Name, elemType); err != nil {
```
why can't we just call generateMapProperty here?

jbrette replied:
@stts done

sttts commented on pkg/generators/openapi.go:
```diff
@@ -671,7 +698,9 @@ func (g openAPITypeWriter) generateSliceProperty(t *types.Type) error {
+		if err := g.generateSliceProperty(elemType); err != nil {
+			return err
+		}
 	default:
```
do we miss the map case here? I.e. arrays of maps.

jbrette replied:
@stts done. Added test for more coverage.

sttts replied:
sttts ;-)

jbrette commented:
@sttts Gentle bump of the PR. Do I need to fix something else ?

sttts commented:
/approve
/lgtm
/hold

@jbrette looks good. Last thing: please create a PR against kube revendoring this. I want to see whether it changes anything in the generated spec.

jbrette commented:
> /approve
> /lgtm
> /hold
> 
> @jbrette looks good. Last thing: please create a PR against kube revendoring this. I want to see whether it changes anything in the generated spec.

I'm not sure I understand what you are asking. Can you sent me a link of a example.

sttts commented:
> I'm not sure I understand what you are asking. Can you sent me a link of a example.

Use `hack/pin-dependency.sh` and `hack/update-vendor.sh` to get your changed kube-openapi into github.com/kubernetes/kubernetes. Then regenerate OpenAPI with the hack/ script and paste the difference here (at best, there is none).

jbrette commented:
@sttts  done. According to ./hack/verify-openapi-spec.sh there is no difference. Neither the openapi.generated.go nor the swagger.json have changed.

Since this is the first time, I'm going through the procedure, here are the steps I followed:

in keleustes fork of kube-openapi
```bash
git checkout mapmap
git tag -a "v1.2.3" -m "v1.2.3"
git push origin v1.2.3
```

Update go.mod
```bash
k8s.io/kube-openapi => github.com/keleustes/kube-openapi v1.2.3
```

run hack-vendor.sh
```bash
hack/update-vendor.sh
```


```bash
xxxx@airship:~/src/k8s.io/kubernetes$ git status
On branch master
Your branch is up-to-date with 'origin/master'.
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

        modified:   go.mod
        modified:   go.sum
        modified:   staging/src/k8s.io/api/go.mod
        modified:   staging/src/k8s.io/api/go.sum
        modified:   staging/src/k8s.io/apiextensions-apiserver/go.mod
        modified:   staging/src/k8s.io/apiextensions-apiserver/go.sum
        modified:   staging/src/k8s.io/apimachinery/go.mod
        modified:   staging/src/k8s.io/apimachinery/go.sum
        modified:   staging/src/k8s.io/apiserver/go.mod
        modified:   staging/src/k8s.io/apiserver/go.sum
        modified:   staging/src/k8s.io/cli-runtime/go.mod
        modified:   staging/src/k8s.io/cli-runtime/go.sum
        modified:   staging/src/k8s.io/client-go/go.mod
        modified:   staging/src/k8s.io/client-go/go.sum
        modified:   staging/src/k8s.io/cloud-provider/go.mod
        modified:   staging/src/k8s.io/cloud-provider/go.sum
        modified:   staging/src/k8s.io/cluster-bootstrap/go.mod
        modified:   staging/src/k8s.io/cluster-bootstrap/go.sum
        modified:   staging/src/k8s.io/code-generator/go.mod
        modified:   staging/src/k8s.io/code-generator/go.sum
        modified:   staging/src/k8s.io/component-base/go.mod
        modified:   staging/src/k8s.io/component-base/go.sum
        modified:   staging/src/k8s.io/csi-translation-lib/go.mod
        modified:   staging/src/k8s.io/csi-translation-lib/go.sum
        modified:   staging/src/k8s.io/kube-aggregator/go.mod
        modified:   staging/src/k8s.io/kube-aggregator/go.sum
        modified:   staging/src/k8s.io/kube-controller-manager/go.mod
        modified:   staging/src/k8s.io/kube-controller-manager/go.sum
        modified:   staging/src/k8s.io/kube-proxy/go.mod
        modified:   staging/src/k8s.io/kube-proxy/go.sum
        modified:   staging/src/k8s.io/kube-scheduler/go.mod
        modified:   staging/src/k8s.io/kube-scheduler/go.sum
        modified:   staging/src/k8s.io/kubectl/go.mod
        modified:   staging/src/k8s.io/kubectl/go.sum
        modified:   staging/src/k8s.io/kubelet/go.mod
        modified:   staging/src/k8s.io/kubelet/go.sum
        modified:   staging/src/k8s.io/legacy-cloud-providers/go.mod
        modified:   staging/src/k8s.io/legacy-cloud-providers/go.sum
        modified:   staging/src/k8s.io/metrics/go.mod
        modified:   staging/src/k8s.io/metrics/go.sum
        modified:   staging/src/k8s.io/node-api/go.mod
        modified:   staging/src/k8s.io/node-api/go.sum
        modified:   staging/src/k8s.io/sample-apiserver/go.mod
        modified:   staging/src/k8s.io/sample-apiserver/go.sum
        modified:   staging/src/k8s.io/sample-cli-plugin/go.mod
        modified:   staging/src/k8s.io/sample-cli-plugin/go.sum
        modified:   staging/src/k8s.io/sample-controller/go.mod
        modified:   staging/src/k8s.io/sample-controller/go.sum
        modified:   vendor/k8s.io/kube-openapi/pkg/generators/openapi.go
        modified:   vendor/modules.txt
```

the openapi.go contains the change we are about to merge

```go
func (g openAPITypeWriter) generateMapProperty(t *types.Type) error {
        keyType := resolveAliasAndPtrType(t.Key)
        elemType := resolveAliasAndPtrType(t.Elem)

        // According to OpenAPI examples, only map from string is supported
        if keyType.Name.Name != "string" {
                return fmt.Errorf("map with non-string keys are not supported by OpenAPI in %v", t)
        }
        g.Do("Type: []string{\"object\"},\n", nil)
        g.Do("AdditionalProperties: &spec.SchemaOrBool{\nAllows: true,\nSchema: &spec.Schema{\nSchemaProps: spec.SchemaProps{\n", nil)
        typeString, format := openapi.GetOpenAPITypeFormat(elemType.String())
        if typeString != "" {
                g.generateSimpleProperty(typeString, format)
                g.Do("},\n},\n},\n", nil)
                return nil
        }
        switch elemType.Kind {
        case types.Builtin:
                return fmt.Errorf("please add type %v to getOpenAPITypeFormat function", elemType)
        case types.Struct:
                g.generateReferenceProperty(elemType)
        case types.Slice, types.Array:
                if err := g.generateSliceProperty(elemType); err != nil {
                        return err
                }
        case types.Map:
                if err := g.generateMapProperty(elemType); err != nil {
                        return err
                }
        default:
                return fmt.Errorf("map Element kind %v is not supported in %v", elemType.Kind, t.Name)
        }
        g.Do("},\n},\n},\n", nil)
        return nil
}
```
the go.mod and go.sum have changed:

```bash
git status | grep modified | awk '{print $2}' > changedList
xxxx@airship:~/src/k8s.io/kubernetes$ grep keleustes `cat changedList`
go.mod: k8s.io/kube-openapi => github.com/keleustes/kube-openapi v1.2.3
go.sum:github.com/keleustes/kube-openapi v1.2.3 h1:GkB0SaZIDmrIinoGxoPkIWYXKXN373b/22o6pdJpBvU=
go.sum:github.com/keleustes/kube-openapi v1.2.3/go.mod h1:1TqjTSzOxsLGIKfj0lK8EeCP7K1iUG65v09OM0/WG5E=
staging/src/k8s.io/api/go.mod:  k8s.io/kube-openapi => github.com/keleustes/kube-openapi v1.2.3
staging/src/k8s.io/api/go.sum:github.com/keleustes/kube-openapi v1.2.3/go.mod h1:1TqjTSzOxsLGIKfj0lK8EeCP7K1iUG65v09OM0/WG5E=
staging/src/k8s.io/apiextensions-apiserver/go.mod:      k8s.io/kube-openapi => github.com/keleustes/kube-openapi v1.2.3
staging/src/k8s.io/apiextensions-apiserver/go.sum:github.com/keleustes/kube-openapi v1.2.3 h1:GkB0SaZIDmrIinoGxoPkIWYXKXN373b/22o6pdJpBvU=
staging/src/k8s.io/apiextensions-apiserver/go.sum:github.com/keleustes/kube-openapi v1.2.3/go.mod h1:1TqjTSzOxsLGIKfj0lK8EeCP7K1iUG65v09OM0/WG5E=
staging/src/k8s.io/apimachinery/go.mod: k8s.io/kube-openapi => github.com/keleustes/kube-openapi v1.2.3
staging/src/k8s.io/apimachinery/go.sum:github.com/keleustes/kube-openapi v1.2.3 h1:GkB0SaZIDmrIinoGxoPkIWYXKXN373b/22o6pdJpBvU=
staging/src/k8s.io/apimachinery/go.sum:github.com/keleustes/kube-openapi v1.2.3/go.mod h1:1TqjTSzOxsLGIKfj0lK8EeCP7K1iUG65v09OM0/WG5E=
staging/src/k8s.io/apiserver/go.mod:    k8s.io/kube-openapi => github.com/keleustes/kube-openapi v1.2.3
staging/src/k8s.io/apiserver/go.sum:github.com/keleustes/kube-openapi v1.2.3 h1:GkB0SaZIDmrIinoGxoPkIWYXKXN373b/22o6pdJpBvU=
staging/src/k8s.io/apiserver/go.sum:github.com/keleustes/kube-openapi v1.2.3/go.mod h1:1TqjTSzOxsLGIKfj0lK8EeCP7K1iUG65v09OM0/WG5E=
```

```bash
./hack/verify-openapi-spec.sh
etcd v3.3.15 already installed. To use:
export PATH="/home/xxxx/src/k8s.io/kubernetes/third_party/etcd:${PATH}"
make: Entering directory '/home/xxxx/src/k8s.io/kubernetes'
make[1]: Entering directory '/home/xxxx/src/k8s.io/kubernetes'
stat /home/xxxx/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/third_party/etcd/Documentation/README.md: no such file or directory
make[1]: Leaving directory '/home/xxxx/src/k8s.io/kubernetes'
+++ [0917 14:30:04] Building go targets for linux/amd64:
    cmd/kube-apiserver
make: Leaving directory '/home/xxxx/src/k8s.io/kubernetes'
make: Entering directory '/home/xxxx/src/k8s.io/kubernetes'
make[1]: Entering directory '/home/xxxx/src/k8s.io/kubernetes'
stat /home/xxxx/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/third_party/etcd/Documentation/README.md: no such file or directory
make[1]: Leaving directory '/home/xxxx/src/k8s.io/kubernetes'
+++ [0917 14:30:26] Building go targets for linux/amd64:
    cmd/kube-apiserver
make: Leaving directory '/home/xxxx/src/k8s.io/kubernetes'
etcd --advertise-client-urls http://127.0.0.1:2379 --data-dir /tmp/tmp.y8tBUIXsZA --listen-client-urls http://127.0.0.1:2379 --debug > "/dev/null" 2>/dev/null
Waiting for etcd to come up.
+++ [0917 14:30:42] On try 1, etcd: : http://127.0.0.1:2379
{"action":"set","node":{"key":"/_test","value":"","modifiedIndex":4,"createdIndex":4}}
+++ [0917 14:30:42] Starting kube-apiserver
+++ [0917 14:30:51] On try 10, apiserver: : ok
+++ [0917 14:30:51] Updating
    /home/xxxx/src/k8s.io/kubernetes/hack/../api/openapi-spec
+++ [0917 14:30:52] SUCCESS
+++ [0917 14:30:52] Clean up complete
diffing /home/xxxx/src/k8s.io/kubernetes/api/openapi-spec against freshly generated openapi spec
/home/xxxx/src/k8s.io/kubernetes/api/openapi-spec up to date.
```

**swagger.json are identical**

Preserve the generated go file

**cp pkg/generated/openapi/zz_generated.openapi.go ../preserve.go**

 Reset the branch

```bash
git reset --hard origin/master

xxxx@airship:~/src/k8s.io/kubernetes$ git status
On branch master
Your branch is up-to-date with 'origin/master'.
nothing to commit, working directory clean
```

Regenerate the openapi.generate.go

```bash
xxxx@airship:~/src/k8s.io/kubernetes$ ./hack/verify-openapi-spec.sh
etcd v3.3.15 already installed. To use:
export PATH="/home/xxxx/src/k8s.io/kubernetes/third_party/etcd:${PATH}"
make: Entering directory '/home/xxxx/src/k8s.io/kubernetes'
make[1]: Entering directory '/home/xxxx/src/k8s.io/kubernetes'
stat /home/xxxx/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/third_party/etcd/Documentation/README.md: no such file or directory
+++ [0917 14:33:57] Building go targets for linux/amd64:
    ./vendor/k8s.io/kube-openapi/cmd/openapi-gen
make[1]: Leaving directory '/home/xxxx/src/k8s.io/kubernetes'
+++ [0917 14:34:12] Building go targets for linux/amd64:
    cmd/kube-apiserver
make: Leaving directory '/home/xxxx/src/k8s.io/kubernetes'
make: Entering directory '/home/xxxx/src/k8s.io/kubernetes'
make[1]: Entering directory '/home/xxxx/src/k8s.io/kubernetes'
stat /home/xxxx/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/third_party/etcd/Documentation/README.md: no such file or directory
make[1]: Leaving directory '/home/xxxx/src/k8s.io/kubernetes'
+++ [0917 14:34:31] Building go targets for linux/amd64:
    cmd/kube-apiserver
make: Leaving directory '/home/xxxx/src/k8s.io/kubernetes'
etcd --advertise-client-urls http://127.0.0.1:2379 --data-dir /tmp/tmp.zoWxH77sjS --listen-client-urls http://127.0.0.1:2379 --debug > "/dev/null" 2>/dev/null
Waiting for etcd to come up.
+++ [0917 14:34:47] On try 1, etcd: : http://127.0.0.1:2379
{"action":"set","node":{"key":"/_test","value":"","modifiedIndex":4,"createdIndex":4}}
+++ [0917 14:34:47] Starting kube-apiserver
+++ [0917 14:34:57] On try 10, apiserver: : ok
+++ [0917 14:34:57] Updating
    /home/xxxx/src/k8s.io/kubernetes/hack/../api/openapi-spec
+++ [0917 14:34:57] SUCCESS
+++ [0917 14:34:57] Clean up complete
diffing /home/xxxx/src/k8s.io/kubernetes/api/openapi-spec against freshly generated openapi spec
/home/xxxx/src/k8s.io/kubernetes/api/openapi-spec up to date.
```

**Compare the zzz_generated.openapi.go . No difference.**

```bash
diff ../preserve.go pkg/generated/openapi/zz_generated.openapi.go
```

jbrette commented:
@stts Is everything ok now ?

sttts commented:
/hold cancel
